### PR TITLE
Add concurrency support to DDB client

### PR DESCRIPTION
**Pull request by arjunkumargiri**
### Description
Add concurrency support to DDB client. Following changes has been included:
* Set sequence number in CRUD method response. 
* Set or update sequence number in case of create.
* Increment sequence number during update 
* Set data object under _source field

 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java [hunk 1]
```diff
@@ -321,14 +352,36 @@ private XContentParser createParser(String json) throws IOException {
         return jsonXContent.createParser(NamedXContentRegistry.EMPTY, DeprecationHandler.IGNORE_DEPRECATIONS, json);
     }
 
-    private String simulateOpenSearchResponse(String index, String id, String source, Map<String, Object> additionalFields) {
+    private GetItemRequest getGetItemRequest(String requestTenantId, String documentId, String index) {
```
Comment (dbwiddis) [inline plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java:355]: (Style) `getGet...` looks like a typo even if it's not. How about `buildGet...` ?

File: plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java [hunk 2]
```diff
@@ -108,25 +112,37 @@ public CompletionStage<PutDataObjectResponse> putDataObjectAsync(PutDataObjectRe
...
         final String tableName = getTableName(request.index());
+        final GetItemRequest getItemRequest = getGetItemRequest(request.tenantId(), request.id(), request.index());
+
         return CompletableFuture.supplyAsync(() -> AccessController.doPrivileged((PrivilegedAction<PutDataObjectResponse>) () -> {
             try {
+                GetItemResponse getItemResponse = dynamoDbClient.getItem(getItemRequest);
+                Long sequenceNumber = 1L;
+                if (getItemResponse != null && getItemResponse.item() != null && getItemResponse.item().containsKey(SEQ_NO_KEY)) {
+                    sequenceNumber = Long.parseLong(getItemResponse.item().get(SEQ_NO_KEY).n()) + 1;
+                }
```
Comment (dbwiddis) [inline plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java:123]: (Readability) Might be nice to excerpt these lines into a local private method `initOrIncrementSeqNo()`

File: plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java [hunk 3]
```diff
@@ -108,25 +112,37 @@ public CompletionStage<PutDataObjectResponse> putDataObjectAsync(PutDataObjectRe
         final String id = request.id() != null ? request.id() : UUID.randomUUID().toString();
         final String tenantId = request.tenantId() != null ? request.tenantId() : DEFAULT_TENANT;
         final String tableName = getTableName(request.index());
+        final GetItemRequest getItemRequest = getGetItemRequest(request.tenantId(), request.id(), request.index());
+
         return CompletableFuture.supplyAsync(() -> AccessController.doPrivileged((PrivilegedAction<PutDataObjectResponse>) () -> {
             try {
+                GetItemResponse getItemResponse = dynamoDbClient.getItem(getItemRequest);
+                Long sequenceNumber = 1L;
```
Comment (dbwiddis) [inline plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java:120]: I think the initial sequence number is supposed to be 0.

File: plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java [hunk 4]
```diff
@@ -142,33 +158,34 @@ public CompletionStage<PutDataObjectResponse> putDataObjectAsync(PutDataObjectRe
...
                 if (getItemResponse == null || getItemResponse.item() == null || getItemResponse.item().isEmpty()) {
                     found = false;
                     sourceObject = null;
                 } else {
                     found = true;
-                    sourceObject = JsonTransformer.convertDDBAttributeValueMapToObjectNode(getItemResponse.item());
+                    sourceObject = JsonTransformer.convertDDBAttributeValueMapToObjectNode(getItemResponse.item().get(SOURCE).m());
+                    if (getItemResponse.item().containsKey(SEQ_NO_KEY)) {
+                        sequenceNumberString = getItemResponse.item().get(SEQ_NO_KEY).n();
+                    }
```
Comment (dbwiddis) [inline plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java:176]: Shouldn't there always be a `SEQ_NO_KEY` if the object exists?  We should put 0 on initial indexing.

File: plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java [hunk 5]
```diff
@@ -219,20 +236,34 @@ public CompletionStage<UpdateDataObjectResponse> updateDataObjectAsync(UpdateDat
                     .tableName(getTableName(request.index()))
                     .key(updateKey)
                     .attributeUpdates(updateAttributeValue);
+                updateItemRequestBuilder
+                    .updateExpression("SET #seqNo = #seqNo + :incr")
+                    .expressionAttributeNames(Map.of("#seqNo", SEQ_NO_KEY))
+                    .expressionAttributeValues(Map.of(":incr", AttributeValue.builder().n("1").build()));
```
Comment (dbwiddis) [inline plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java:242]: Again, pretty sure this should be 0.  See [here](https://github.com/opensearch-project/OpenSearch/blob/77a74e2fcf2c6b56b0959a563fbc6a0d7aff220a/server/src/main/java/org/opensearch/index/seqno/SequenceNumbers.java#L52-L55); `NO_OPS_PERFORMED` is -1 so the initial indexing operation increments to 0.

File: plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java [hunk 6]
```diff
@@ -273,16 +304,16 @@ public CompletionStage<DeleteDataObjectResponse> deleteDataObjectAsync(DeleteDat
...
-                // TODO need to return SEQ_NO here
-                // If document doesn't exist, increment and return highest seq no ever seen, but we would have to track seqNo here
-                // If document never existed, return -2 (unassigned) for seq no (probably what we have to do here)
-                // If document exists, increment and return SEQ_NO
-                dynamoDbClient.deleteItem(deleteItemRequest);
-                // TODO need to pass seqNo to simulated response
+                DeleteItemResponse deleteItemResponse = dynamoDbClient.deleteItem(deleteItemRequest);
+                Long sequenceNumber = null;
+                if (deleteItemResponse.attributes() != null && deleteItemResponse.attributes().containsKey(SEQ_NO_KEY)) {
+                    sequenceNumber = Long.parseLong(deleteItemResponse.attributes().get(SEQ_NO_KEY).n());
```
Comment (dbwiddis) [inline plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java:310]: The seqNo should be incremented on a deletion; yes, it's gone and it'll end up `-2` later, but any existing testing will expect a delete response to increment this.

File: plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java [hunk 7]
```diff
@@ -321,14 +352,36 @@ private XContentParser createParser(String json) throws IOException {
...
+    ) {
         StringBuilder sb = new StringBuilder("{");
         // Fields with a DDB counterpart
         sb.append("\"_index\":\"").append(index).append("\",");
         sb.append("\"_id\":\"").append(id).append("\",");
         // Fields we must simulate using default values
-        sb.append("\"_primary_term\":").append(UNASSIGNED_PRIMARY_TERM).append(",");
-        sb.append("\"_seq_no\":").append(UNASSIGNED_SEQ_NO).append(",");
+        sb.append("\"_primary_term\":").append(sequenceNumber == null ? UNASSIGNED_PRIMARY_TERM : 1).append(",");
+        sb.append("\"_seq_no\":").append(sequenceNumber == null ? UNASSIGNED_SEQ_NO : sequenceNumber).append(",");
```
Comment (dbwiddis) [inline plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java:384]: (Readability) Might be cleaner to wrap these two lines in an `if (sequenceNumber == null)` conditional to make the pairing more clear (i.e., unassigned for both, or seqNo >=0 + primaryTerm=1)

File: plugin/src/test/java/org/opensearch/ml/sdkclient/DDBOpenSearchClientTests.java [hunk 8]
```diff
@@ -151,17 +153,44 @@ public void testPutDataObject_HappyCase() throws IOException {
         IndexResponse indexActionResponse = IndexResponse.fromXContent(response.parser());
         assertEquals(TEST_ID, indexActionResponse.getId());
         assertEquals(DocWriteResponse.Result.CREATED, indexActionResponse.getResult());
+        assertEquals(1, indexActionResponse.getSeqNo());
```
Comment (dbwiddis) [inline plugin/src/test/java/org/opensearch/ml/sdkclient/DDBOpenSearchClientTests.java:156]: Should be 0 :)

**Review by dbwiddis — COMMENTED** [review]
Generally LGTM.
 - some style/readability suggestions
 - seqNo should be initialized to 0 on a new document
 - seqNo should be incremented on a deletion (even if we don't save that # anywhere)

File: plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java (hunk 4 above)
Comment (arjunkumargiri) [reply plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java:176]: Yes, `SEQ_NO_KEY` will be set always be set when a new DDB entry is created. However, having contains check for code completeness.

File: plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java (hunk 6 above)
Comment (arjunkumargiri) [reply plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java:310]: Got it, increase seq_no in response object.

File: plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java (hunk 5 above)
Comment (dbwiddis) [reply plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java:240]: Ah, got it, this line of code says "increment by one".  

Would have helped with the classic needless comment:
```java
// increment n by one
n++;
```
😆

**Review by dbwiddis — APPROVED** [review]
LGTM!

**Review by dhrubo-os — APPROVED** [review]

**Pull request merged**